### PR TITLE
Persist recently used blocks, and record block usage

### PR DESCRIPTION
**Pull request by notnownikki**
## Description

Loading up a new post populated the recently used blocks with
blocks from the common category. This change persists the
recently used blocks, so they are there next time.

This also keeps a "highscore table" of block usage, so in
a future PR, we can use this data to display the user's
most frequently used blocks next to the inserter, where we
currently hardcode paragraph and image. The work to do
that will be in a future PR so that when it's released, we
have a better chance of the user having usage data that
is meaningful to base it on.

## How Has This Been Tested?

Add some blocks to a post, they should appear at the
top of the recently used blocks.

Start a new post, or load an existing post, and the blocks in
the recent tab should show the blocks you used most
recently.

## Types of changes
New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: editor/reducer.js
Comment (youknowriad) [inline editor/reducer.js]: Is this line necessary since we're overriding the value just after?

File: editor/reducer.js
Comment (youknowriad) [inline editor/reducer.js]: Why do we need to store `blockUsage` in state and do this initialization of `recentlyUsedBlocks` since `recentlyUsedBlocks` is also persisted?

Comment (notnownikki) [reply editor/reducer.js]: I wanted some feedback on this... when we start a new post, we want to populate the recently used blocks based on what blocks the user uses most frequently, so their most common stuff is there ready for them, instead of perhaps the twitter embed they finished their last post with, if you see what I mean. So, the initial state of `recentlyUsedBlocks` relies on `blockUsage`, but I couldn't get to `blockUsage` from a separate reducer, so it moved into preferences. Open to suggestions here!

Comment (youknowriad) [reply editor/reducer.js]: I don't have a strong preference on the algorithm itself, but in this case, I'd assume that `INSERT_BLOCKS` works similarly. If the newly created block is not among the most used blocks, don't add it. I'm thinking we should be consistent.

Also, when do we start counting the usage, should we fix a threshold or keep these numbers grow infinitely?

Comment (notnownikki) [reply editor/reducer.js]: > If the newly created block is not among the most used blocks, don't add it.

Then how do we get any most used blocks? We'd never be able to add any, or they'd never be able to change. Or do you mean that if a block is already in the most used, don't increment?

Either way, we're just recording an int, so if I remember correctly, we can record up 
 to `Number.MAX_SAFE_INTEGER` which is 9007199254740991. I'm not sure most of our users will ever insert that many of a single block?

File: editor/reducer.js
Comment (notnownikki) [reply editor/reducer.js]: I thought this is part of the principle of always creating a new values, rather than changing existing data. The state should be `deepFreeze`'d here in tests, so trying to change `blockUsage` should result in an error, because we'd be modifying the state directly.

File: editor/reducer.js
Comment (notnownikki) [reply editor/reducer.js]: I'm still not sure how that would work, because we want to record how frequently blocks are used so that we can generate a list of the most frequently used blocks. If we didn't increment the usage, then we wouldn't end up with useful data... or am I understanding this wrong? I'm not sure what you're getting at here.

Comment (notnownikki) [reply editor/reducer.js]: Perhaps I'm not explaining the problem I'm solving very well..

I want this to do three things:

1) Keep track of what blocks a user uses most frequently.
2) Use the usage data to populate the "recent" tab with their most frequently used blocks on editor setup
3) During an editing session, have the most recently used blocks appear in the "recent" tab.

As far as I can see, that requires we keep track of how many times a user uses blocks, and either have a list of recently used ones, or keep `last_used` data. Personally, I'd rather go for the latter, as two data structures seems wrong when it's really just two aspects of block usage, but as you want to have a memoized selector based on that data, that becomes difficult. Why do you want to memoize? Is it to avoid getting the concrete blocks each time?

Comment (youknowriad) [reply editor/reducer.js]: I didn't say don't increment the usage :). I just use one unique technique: 

 - increment usage on insert and use the selector above
 - or do not save the usage at all and rely on the last inserted blocks (like before)

Comment (youknowriad) [reply editor/reducer.js]: So what you're suggesting is that the user have a list of rencent blocks shown but when you refresh the page, this list changes completely because we'll rely on the usage to recompute this list. Why this difference? I just think we should go for consistency and keep the same list when refreshing.

Comment (notnownikki) [reply editor/reducer.js]: Yes, that's what this does, and why I wanted feedback on it :smile: 

Perhaps we do just want to persist the recently used list itself? If someone uses a block frequently, it's unlikely to drop out of the list, although it may change position, as we put the most recently used ones at the top. Or perhaps it's always a list of the most frequently used blocks, ordered by how often they're used?

Comment (youknowriad) [reply editor/reducer.js]: Yep, I don't know the answer. My suggestion was being consistent across reloads but let's what others think @karmatosed @jasmussen

**Comment by jasmussen** [comment]
Really nice work, thanks for working on this.

Took the branch for a spin, did this:

- Looked at inserter, saw paragraph first
- Inserted Latest Posts, saw latest posts sorted at the top

This felt expected, very nice! Then created a new post, saw the Latest Posts block sorted not at first, like it was previously. That was a bit confusing.

> Perhaps we do just want to persist the recently used list itself? If someone uses a block frequently, it's unlikely to drop out of the list, although it may change position, as we put the most recently used ones at the top. Or perhaps it's always a list of the most frequently used blocks, ordered by how often they're used?

This sounds right to me. That is — if I keep using the "Recent Posts" block, it will ride up the "highscore list" all the way to the top. 

The key here is that the recent lists changes per your usage, but persists so it's "the same" in between posts, and only changes when you actually use other blocks.

Does that make sense? Did I grok this right?

**Comment by notnownikki** [comment]
@jasmussen it's a little bit more subtle...

> This sounds right to me. That is — if I keep using the "Recent Posts" block, it will ride up the "highscore list" all the way to the top.

We can't use the "highscore" by itself. The problem is, if we have 8 blocks we use a lot, then we use one we've never used before, it won't have a score high enough to appear in the list.

If we use the "last used time" by itself, then the list becomes purely a list of what we've used recently, with most recent at the top, which I think is what people are expecting, although the list would change order depending on what blocks you've used most recently. However, this is exactly how a "recently opened documents" list works, so perhaps what people expect.

From the feedback I'm getting here, it seems like the "highscore" isn't of any use, and we should just be persisting the list based on when blocks were last used.

**Comment by jasmussen** [comment]
It's a bit above my head, so I would defer to Riad here — but just wanted to note your work here is really appreciated, thank you.

**Comment by notnownikki** [comment]
I really think that we just persist the recently used state. Mixing the highscore table and the recently used state is just confusing.

HOWEVER, I think it's still worth keeping the highscore table (adopting that term, I like it haha) because the quick links next to the inserter icon (currently for paragraph and image) could be generated by the blocks a user most frequently uses, but that can come in a future PR.

**Comment by notnownikki** [comment]
I've changed this around now so it only uses the blocks you have used recently for the recently used blocks tab, displaying them with most recent at the top.

The block usage highscore is still in there, once we're happy with this branch and it's merged, I'll do a separate PR for the frequently used blocks to appear next to the inserter, where paragraph and image currently are.

File: editor/reducer.js
Comment (youknowriad) [inline editor/reducer.js]: What if some recently used blocks do not exist any more, should we filter the list here or is this done somewhere else?

Comment (notnownikki) [reply editor/reducer.js]: Ah, good catch! I'll look at that now.

Comment (notnownikki) [reply editor/reducer.js]: I think the filtering should probably be done on editor setup, so we can filter any stored block information in the same step.

Comment (youknowriad) [reply editor/reducer.js]: don't you think the  `SETUP_EDITOR` is a good place for this (it's editor setup step) or are you thinking fo somewhere else?

What concerns me in moving this behaviour elsewhere is having to know the shape of the store outside the store.

Comment (notnownikki) [reply editor/reducer.js]: I thought `SETUP_EDITOR` would be the place to do it, in the `preferences` reducer, so we can filter out (or maybe migrate in future?) any block data that's in preferences, and should no longer be there.

Comment (notnownikki) [reply editor/reducer.js]: Fixed! Couple of filter functions in the preferences reducer that remove invalid block names from a list, or invalid block names from an object's properties.

**Comment by notnownikki** [comment]
Alrighty, I have updated the title and description here to reflect the outcome of our discussions. Thanks for all the reviews so far! Hopefully this makes sense now, and we can get it out there and start populating the highscore tables :)

File: editor/reducer.js
Comment (youknowriad) [inline editor/reducer.js]: Minor: This could be factorized `const isBlockDefined = name => getBlockType( name ) !== undefined` and reused in both `filterInvalidBlocksFromList` and `filterInvalidBlocksFromObject`
I don't know if it makes it clearer though

File: editor/reducer.js
Comment (youknowriad) [inline editor/reducer.js]: Minor: Could this be defined as a module constant above: `const MAX_RECENT_BLOCKS = 8;` (we use this unofficial guideline elsewhere for constants)

**Review by youknowriad — APPROVED** [review]
Good job 👍

File: editor/reducer.js
Comment (notnownikki) [reply editor/reducer.js]: I like it, certainly makes it explicit what is being checked for, instead of duplicating the logic of how we check. Fixed!

File: editor/reducer.js
Comment (youknowriad) [inline editor/reducer.js]: you can use `pick` to avoid negating `isBlockDefined` :)

**Pull request merged**
